### PR TITLE
feat: Enable columnar shuffle by default

### LINKED ISSUE #95: Enable columnar shuffle by default
### What is the problem the feature request solves?

`COMET_COLUMNAR_SHUFFLE_ENABLED` is the config Comet uses to decide if columnar shuffle should be used instead of native shuffle, if comet shuffle is enabled.

Because columnar shuffle covers more use cases (hash, range and single partitioning) than native shuffle. We should change this config to prefer columnar shuffle.

Note that this is blocked by the Java Arrow issue: https://github.com/apache/arrow/issues/40038 (PR: https://github.com/apache/arrow/pull/40043)

Currently if columnar shuffle is enabled, running Comet with TPCDS queries will get the error:

```
General execution error with reason org.apache.comet.CometNativeException: Fail to process Arrow array with reason C Data interface error: The external buffer at position 1 is null...
```

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #95.

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/CometTPCDSQuerySuite.scala [hunk 1]
```diff
@@ -154,8 +154,9 @@ class CometTPCDSQuerySuite
     conf.set(CometConf.COMET_EXEC_ALL_OPERATOR_ENABLED.key, "true")
     conf.set(CometConf.COMET_EXEC_ALL_EXPR_ENABLED.key, "true")
     conf.set(CometConf.COMET_EXEC_SHUFFLE_ENABLED.key, "true")
+    conf.set(CometConf.COMET_MEMORY_OVERHEAD.key, "20g")
     conf.set(MEMORY_OFFHEAP_ENABLED.key, "true")
-    conf.set(MEMORY_OFFHEAP_SIZE.key, "2g")
```
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/test/scala/org/apache/spark/sql/CometTPCDSQuerySuite.scala:158]: Maybe we don't have enough explanation for these configurations in developer documentation so far. We should update the necessary documentation later.

**Comment by viirya** [comment]
Hmm, all tests in TPCDSQuerySuite` are passed locally with this PR. I need to look at the CI failure.

**Comment by viirya** [comment]
Observed several Spark SQL test failures regarding aggregation: #260

File: dev/diffs/3.4.2.diff [hunk 2]
```diff
@@ -1414,6 +1424,7 @@ index ed2e309fa07..4cfe0093da7 100644
 +          .set("spark.shuffle.manager",
 +            "org.apache.spark.sql.comet.execution.shuffle.CometShuffleManager")
 +          .set("spark.comet.exec.shuffle.enabled", "true")
++          .set("spark.comet.memoryOverhead", "10g")
```
Comment (viirya) [inline dev/diffs/3.4.2.diff:1427]: Observed that Comet is unable to acquire enough memory for columnar shuffle when doing Spark SQL tests:

For example, DatasetPrimitiveSuite:

```
  Cause: org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 37.0 failed 1 times, most recent failure: Lost task 1.0 in stage 37.0 (TID 75) (e4773b5abe7e executor driver): org.apache.spark.memory.SparkOutOfMemoryError: Unable to acquire 67108848 bytes of memory, got 96 bytes. Available: 96
[info] 	at org.apache.spark.shuffle.comet.CometShuffleMemoryAllocator.allocate(CometShuffleMemoryAllocator.java:132)
[info] 	at org.apache.spark.shuffle.comet.CometShuffleMemoryAllocator.allocatePage(CometShuffleMemoryAllocator.java:119)
[info] 	at org.apache.spark.sql.comet.execution.shuffle.SpillWriter.initialCurrentPage(SpillWriter.java:158)
[info] 	at org.apache.spark.sql.comet.execution.shuffle.CometDiskBlockWriter.insertRow(CometDiskBlockWriter.java:284)
```

Comment (viirya) [reply dev/diffs/3.4.2.diff:1427]: Increased Comet memoryoverhead to overcome it.

**Comment by viirya** [comment]
I fixed all Spark SQL test failures. Now waiting for #380 to be merged.

**Comment by viirya** [comment]
cc @sunchao @andygrove

File: spark/src/main/scala/org/apache/spark/sql/comet/plans/AliasAwareOutputExpression.scala [hunk 3]
```diff
@@ -31,7 +31,8 @@ import org.apache.spark.sql.catalyst.trees.CurrentOrigin
 trait AliasAwareOutputExpression extends SQLConfHelper {
   // `SQLConf.EXPRESSION_PROJECTION_CANDIDATE_LIMIT` is Spark 3.4+ only.
   // Use a default value for now.
-  protected val aliasCandidateLimit = 100
+  protected val aliasCandidateLimit: Int =
+    conf.getConfString("spark.sql.optimizer.expressionProjectionCandidateLimit", "100").toInt
```
Comment (viirya) [inline spark/src/main/scala/org/apache/spark/sql/comet/plans/AliasAwareOutputExpression.scala:35]: Some Spark tests tune this config. So we need to get the configured value.

**Comment by viirya** [comment]
@andygrove @sunchao Could you take a look and see if you have any comments on this? Thanks.

File: core/src/execution/datafusion/planner.rs [hunk 4]
```diff
@@ -1039,6 +1039,21 @@ impl PhysicalPlanner {
...
             let full_schema = Arc::new(Schema::new(all_fields));
 
+            // Because we cast dictionary array to array in scan operator,
+            // we need to change dictionary type to data type for join filter expression.
+            let fields: Vec<_> = full_schema
+                .fields()
+                .iter()
+                .map(|f| match f.data_type() {
+                    DataType::Dictionary(_, val_type) => {
+                        Arc::new(Field::new(f.name(), val_type.as_ref().clone(), true))
```
Comment (andygrove) [inline core/src/execution/datafusion/planner.rs:1049]: Should we preserve the fields nullable flag here?

```suggestion
                        Arc::new(Field::new(f.name(), val_type.as_ref().clone(), f.is_nullable()))
```

Comment (viirya) [reply core/src/execution/datafusion/planner.rs:1049]: Yea

File: core/src/execution/datafusion/planner.rs [hunk 5]
```diff
@@ -1057,6 +1072,14 @@ impl PhysicalPlanner {
                         .into_iter()
                         .map(|i| right.schema().field(i).clone()),
                 )
+                // Because we cast dictionary array to array in scan operator,
+                // we need to change dictionary type to data type for join filter expression.
+                .map(|f| match f.data_type() {
+                    DataType::Dictionary(_, val_type) => {
+                        Field::new(f.name(), val_type.as_ref().clone(), true)
```
Comment (andygrove) [inline core/src/execution/datafusion/planner.rs:1079]:
```suggestion
                        Field::new(f.name(), val_type.as_ref().clone(), f.is_nullable())
```

File: spark/src/test/scala/org/apache/spark/sql/CometTPCDSQuerySuite.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (andygrove) [reply spark/src/test/scala/org/apache/spark/sql/CometTPCDSQuerySuite.scala:156]: We do have documentation for these memory configs now. 

https://datafusion.apache.org/comet/user-guide/configs.html

File: dev/diffs/3.4.2.diff [hunk 6]
```diff
@@ -210,6 +210,48 @@ index 0efe0877e9b..423d3b3d76d 100644
...
+ import org.apache.spark.sql.execution.{ColumnarToRowExec, ExecSubqueryExpression, RDDScanExec, SparkPlan}
+ import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanHelper
+ import org.apache.spark.sql.execution.columnar._
+@@ -516,7 +517,8 @@ class CachedTableSuite extends QueryTest with SQLTestUtils
+    */
+   private def verifyNumExchanges(df: DataFrame, expected: Int): Unit = {
+     assert(
+-      collect(df.queryExecution.executedPlan) { case e: ShuffleExchangeExec => e }.size == expected)
++      collect(df.queryExecution.executedPlan) {
++        case _: ShuffleExchangeExec | _: CometShuffleExchangeExec => 1 }.size == expected)
```
Comment (andygrove) [inline dev/diffs/3.4.2.diff:231]: Could we just check for `ShuffleExchangeLike` instead (and even push that change upstream?)

File: spark/src/test/scala/org/apache/spark/sql/CometTPCDSQuerySuite.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (viirya) [reply spark/src/test/scala/org/apache/spark/sql/CometTPCDSQuerySuite.scala:156]: Yea, that's really helpful.

File: dev/diffs/3.4.2.diff (hunk 6 above)
Comment (viirya) [reply dev/diffs/3.4.2.diff:231]: We could. For upstream, not sure if it is accepted, but we can try.

Comment (viirya) [reply dev/diffs/3.4.2.diff]: I changed to check `ShuffleExchangeLike` instead for the places which is possible.

**Comment by viirya** [comment]
Thank you @andygrove

**Comment by viirya** [comment]
Merged. Thanks @andygrove for review.

**Pull request merged**